Opened files now start at first non space character

CajetanP opened the pull request:
I was just reading the editor's code when I noticed that TODO in a buffer.c file, so I decided to take care of it. I think it works pretty well (or at least acceptable) right now.

lpan commented:
I will take a look at it later tonight :+1:

lpan commented on src/buffer.c:
```diff
@@ -124,11 +124,26 @@ buffer_t *init_buffer(const char *filename) {
+  // go to the first non-empty row
+  if (buf->num_rows > 1) {
+      while (buf->current->line_size == 0) {
+          buf->current = buf->current->next;
```
This will potentially cause a `Segmentation fault`. `buf->current` will point to `NULL` if the file consists of three empty lines (line_size is 0).

lpan commented on src/buffer.c:
```diff
@@ -124,11 +124,26 @@ buffer_t *init_buffer(const char *filename) {
+  }
+
+  // go to the first non-empty character
+  if (buf->current->line_size > 1) {
```
Ha! Segmentation fault!

lpan commented on src/buffer.c:
```diff
@@ -124,11 +124,26 @@ buffer_t *init_buffer(const char *filename) {
+  if (buf->num_rows > 1) {
+      while (buf->current->line_size == 0) {
+          buf->current = buf->current->next;
+          buf->current_row++;
```
I like to have a space before `++`

lpan commented on src/buffer.c:
```diff
@@ -124,11 +124,26 @@ buffer_t *init_buffer(const char *filename) {
+  if (buf->current->line_size > 1) {
+      while (buf->current->current->c == ' ') {
+          buf->current->current = buf->current->current->next;
+          buf->current_char++;
```
Space before `++` please

lpan commented:
@CajetanP My finals are coming up in two weeks. After that I am planning to do a big revamp. It is gonna have more rigorous tests, better design. Probably I am going to get rid of NCurses as well.

CajetanP commented:
I fixed the identation & segfaults. I also added noraw() just before ncurses exits because I noticed that my terminal sometimes behaves in a strange way after editor crashes (or exits) and I suspect it may be related to raw mode not being disabled properly.

lpan commented on src/main.c:
```diff
@@ -26,6 +26,7 @@ int main(int argc, char **argv) {
   start_listener(st);
 
   destroy_state(st);
+  noraw();
```
I think this is irrelevant. `raw()` only overwrites ctrl-c, ctrl-z, etc. Even if it somehow fixes the issue, when the editor crashes because of a runtime error, it won't be executed anyway.

lpan commented on src/buffer.c:
```diff
@@ -124,11 +124,26 @@ buffer_t *init_buffer(const char *filename) {
   buf->current_char = 0;
+  buf->current_row = 0;
+
+  // go to the first non-empty row
```
I don't think Vim has such behaviour as it goes to the first non empty char but not first non empty line (if the first line is empty the cursor stays there)

CajetanP commented:
Well I don't really have time to dive into how these tests work so I'm not planning on writing them right now. I guess if it works then it works and we can add thorough tests some time in the future.

lpan commented:
@CajetanP totally agree. LMK if you are ready to merge

CajetanP commented:
Well if you think it's okay then go ahead, it's your project after all.

lpan commented:
@CajetanP I asked just in case if you have something else to add :smile: anyway, thanks for contributing

CajetanP commented:
Sure, I'm happy to help :smile:
